Platform/CommonBoardPkg/Library/TccLib/TccLib.c

```diff
@@ -0,0 +1,639 @@
+
+  @param[in] EFI_ACPI_DESCRIPTION_HEADER   PTCT ACPI Table entry
+
+  @retval EFI_STATUS            Status returned from PTCT Table entries
+          EFI_NOT_FOUND         Loader Global Pointer not found
+                                TccHob not found
+          EFI_INVALID_PARAMETER Incorrect Ptct table entries
+**/
+EFI_STATUS
+EFIAPI
+UpdateAcpiPtctTable (
+    IN EFI_ACPI_DESCRIPTION_HEADER *PtctTable
+  )
+{
+  UINT8                         *DestAddr;
+  EFI_STATUS                    Status;
+  PTCT_HEADER_ENTRY             **PtctList;
+  PTCT_HEADER_ENTRY             *PtctEntry;
+  UINT32                        PtctTableLength;
+  UINT32                        PtctCallTableCount;
+  UINT32                        Index;
+  UINT32                        MaxPtctEntries;
+  UINT32                        PtctEntryListIndex;
+  UINT32                        PtctCallIndex;
+  LOADER_GLOBAL_DATA            *LdrGlobal;
+
+  Status = EFI_SUCCESS;
+  if (PtctTable == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  LdrGlobal   = (LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer();
+  if (LdrGlobal == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  mTccInfoHob = GetNextGuidHob (&gTccInfoHobGuid, LdrGlobal->FspHobList);
+  if (mTccInfoHob == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  PtctTableLength = sizeof (EFI_ACPI_DESCRIPTION_HEADER);
+  PtctEntry = NULL;
+
+  PtctCallTableCount = sizeof (mPtctCallTable)/sizeof (mPtctCallTable[0]);
+
+  MaxPtctEntries = 0;
+
+  for (Index = 0; Index < PtctCallTableCount; Index++) {
+    MaxPtctEntries += mPtctCallTable[Index].NumEntries;
+  }
+
+  PtctList = AllocatePool(MaxPtctEntries * sizeof (PTCT_HEADER_ENTRY*));
+  if (PtctList == NULL) {
+    DEBUG ((DEBUG_ERROR, "PTCT installation failed: out of resources\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+  ZeroMem (PtctList, (MaxPtctEntries*sizeof (PTCT_HEADER_ENTRY*)));
+  PtctEntryListIndex = 0;
+
+  // Get all PTCT Entries
+  for (Index = 0; Index < PtctCallTableCount; Index++) {
+    for (PtctCallIndex = 0; PtctCallIndex < mPtctCallTable[Index].NumEntries; PtctCallIndex++) {
+      Status = mPtctCallTable[Index].ptct_entry_func (PtctCallIndex, &PtctEntry);
+      if (Status == EFI_SUCCESS) {
+        DEBUG ((DEBUG_INFO, "PTCT Entry %d installed\n", Index));
+        PtctList[PtctEntryListIndex] = PtctEntry;
+        PtctEntry = NULL;
+        PtctTableLength += PtctList[PtctEntryListIndex]->Size;
+        PtctEntryListIndex++;
+      }
+    }
+  }
+  DEBUG ((DEBUG_INFO, "%d PTCT entries installed\n", PtctEntryListIndex));
+  PtctTable->Length          = PtctTableLength;
+
+  // Add PTCT entries to ACPI table
+  DestAddr = (UINT8 *) PtctTable;
+  DestAddr += sizeof (EFI_ACPI_DESCRIPTION_HEADER);
+
+  for (Index = 0; Index < MaxPtctEntries; Index++) {
+    if (PtctList[Index] == NULL) {
+      continue;
+    }
+
+  CopyMem (DestAddr, PtctList[Index], PtctList[Index]->Size);
+  DestAddr += PtctList[Index]->Size;

```

Comment:
Can we append necessary entries into PtctTable region directly without temporary memory?